swarm/storage: Pyramid chunker re-write

jmozah opened the pull request:
Complete re-write of pyramid chunker. 
- It doesn't take size now. 
- This is the base for implementing "append"
It can now be used to chunk streams with unknown size.

jmozah commented:
Benchmark result in my machine

zahoor@zahoor.mbp:~/go/src/github.com/ethereum/go-ethereum/swarm/storage (wip-pyramid-chunker) $go test -cpu 4 -bench=. -run no
BenchmarkSplitTree_2-4       	   50000	     29320 ns/op	    3795 B/op	      27 allocs/op
BenchmarkSplitTree_2h-4      	   30000	     60731 ns/op	    4253 B/op	      27 allocs/op
BenchmarkSplitTree_3-4       	   20000	     95937 ns/op	    4773 B/op	      27 allocs/op
BenchmarkSplitTree_3h-4      	    3000	    398683 ns/op	   10928 B/op	      40 allocs/op
BenchmarkSplitTree_4-4       	    2000	    757910 ns/op	   17536 B/op	      46 allocs/op
BenchmarkSplitTree_4h-4      	     500	   3646963 ns/op	   72240 B/op	     106 allocs/op
BenchmarkSplitTree_5-4       	     200	   7207745 ns/op	  140208 B/op	     178 allocs/op
BenchmarkSplitTree_6-4       	      20	  71692609 ns/op	 1370048 B/op	    1512 allocs/op
BenchmarkSplitTree_7-4       	       2	 721174723 ns/op	13666048 B/op	   14820 allocs/op
BenchmarkSplitTree_8-4       	       1	7237853346 ns/op	136621344 B/op	  147871 allocs/op
BenchmarkSplitPyramid_2-4    	  100000	     36174 ns/op	   16221 B/op	      30 allocs/op
BenchmarkSplitPyramid_2h-4   	   30000	     71389 ns/op	   16048 B/op	      30 allocs/op
BenchmarkSplitPyramid_3-4    	   20000	     82921 ns/op	   16494 B/op	      30 allocs/op
BenchmarkSplitPyramid_3h-4   	    5000	    316495 ns/op	   21684 B/op	      37 allocs/op
BenchmarkSplitPyramid_4-4    	    3000	    463458 ns/op	   24468 B/op	      40 allocs/op
BenchmarkSplitPyramid_4h-4   	    1000	   1939678 ns/op	   52308 B/op	      70 allocs/op
BenchmarkSplitPyramid_5-4    	     500	   3705756 ns/op	   85716 B/op	     106 allocs/op
BenchmarkSplitPyramid_6-4    	     100	  36133736 ns/op	  701512 B/op	     775 allocs/op
BenchmarkSplitPyramid_7-4    	       5	 431053261 ns/op	 8240521 B/op	    8922 allocs/op
BenchmarkSplitPyramid_8-4    	     100	4269283181 ns/op	68536216 B/op	   74062 allocs/op
PASS
ok  	github.com/ethereum/go-ethereum/swarm/storage	494.413s

jmozah commented:
BenchmarkAppendPyramid_2-4    	   10000	    176024 ns/op	   54344 B/op	      83 allocs/op
BenchmarkAppendPyramid_2h-4   	   10000	    166854 ns/op	   54342 B/op	      82 allocs/op
BenchmarkAppendPyramid_3-4    	   10000	    193190 ns/op	   55459 B/op	      83 allocs/op
BenchmarkAppendPyramid_4-4    	    5000	    372089 ns/op	   59208 B/op	      84 allocs/op
BenchmarkAppendPyramid_4h-4   	    5000	    382823 ns/op	   59207 B/op	      84 allocs/op
BenchmarkAppendPyramid_5-4    	    5000	    377257 ns/op	   59210 B/op	      84 allocs/op
BenchmarkAppendPyramid_6-4    	    5000	    376271 ns/op	   59206 B/op	      84 allocs/op
BenchmarkAppendPyramid_7-4    	    5000	    405490 ns/op	   59208 B/op	      84 allocs/op
BenchmarkAppendPyramid_8-4    	    5000	    377806 ns/op	   59208 B/op	      84 allocs/op

jmozah commented:
@zelig Added pyramid chunker with BMT.. by mind voice says i have not used your interface for BMT properly.. Also the results i shared with you are too good to believe... please check

jmozah commented:
http://swarm-gateways.net/bzz:/302e7d41eea26e878a7d53cd363170412fa3c5b55925f316d1f495e0743d99a5/

zelig commented on swarm/storage/chunker.go:
```diff
@@ -262,8 +253,11 @@ func (self *TreeChunker) hashWorker(jobC chan *hashJob, chunkC chan *Chunk, errC
+
 	hasher.Write(job.chunk)
 	h := hasher.Sum(nil)
+	hasher.Reset()
```
Reset should come before and the bmt should be fixed to
reserve tree from pool with `Reset()` and release on `Sum()`

zelig commented on swarm/storage/pyramid.go:
```diff
@@ -73,129 +131,486 @@ func NewPyramidChunker(params *ChunkerParams) (self *PyramidChunker) {
+	select {
+	case err := <-errC:
+		if err != nil {
+			close(quitC)
```
now this will close the quitc the second time, so will crash

jmozah replied:
Yeah.. Thanks for catching this.. Fixed it

zelig commented on swarm/storage/pyramid.go:
```diff
@@ -73,129 +131,486 @@ func NewPyramidChunker(params *ChunkerParams) (self *PyramidChunker) {
-	results := Tree{
-		Chunks: chunks,
-		Levels: make([]map[int64]*Node, depth),
+	// waiting for all chunks to finish
```
```golang
go func() {
   wg.Wait()
   if storageWG != nil {
	storageWG.Wait()
  }
  close(errC)
}()
```

and then 

```golang
defer close(quitC)
select {
case err := <-errC:
    if err != nil {
      return nil, err
    }
case <- time.NewTimer(timout).C:
}
return x, nil
```

actually we should probably use context here @lmars any recommendations?

lmars replied:
+1 for using a `context`

jmozah replied:
Can we have it in phase 2? or doe this needs to be in this PR?

zelig commented on swarm/storage/pyramid.go:
```diff
@@ -73,129 +131,486 @@ func NewPyramidChunker(params *ChunkerParams) (self *PyramidChunker) {
+	wg.Add(1)
+	go self.prepareChunks(true, chunkLevel, data, rootKey, quitC, wg, jobC, processorWG, chunkC, errC, storageWG)
+	// waiting for all chunks to finish
+	wg.Wait()
```
same issue as in split, see above

zelig replied:
plus the same error as the one you fixed above with `close(quit)` not alwayscalled

jmozah replied:
Fixed this too

zelig commented on swarm/storage/pyramid.go:
```diff
@@ -73,129 +131,486 @@ func NewPyramidChunker(params *ChunkerParams) (self *PyramidChunker) {
+
+}
+
+func (self *PyramidChunker) enqueTreeChunkAndAddEntryToTree(chunkLevel [][]*TreeEntry, ent *TreeEntry, chunkWG *sync.WaitGroup, jobC chan *chunkJob, quitC chan bool) {
```
"enqueue" please simplify this function name, its a handful :)

jmozah replied:
ha ha... Fixed this too.

jmozah commented:
- Fixed a bug when the chunk ends in multiple of 4096
- Added test cases for TreeChunker Vs BMTChunker correctness test
- Fixed some review changes specified above

zelig commented on bmt/bmt.go:
```diff
@@ -0,0 +1,565 @@
+}
+
+func (self *BMTHasher) Reset() {
+	self.releaseTree()
```
the release should not be here, but in Sum 
There is not much point in releasing here if we immediately reclaim with getTree

zelig commented on bmt/bmt.go:
```diff
@@ -0,0 +1,565 @@
+// hash.Hash interface Sum method appends the byte slice to the underlying
+// data before it calculates and returns the hash of the chunk
+func (self *BMTHasher) Sum(b []byte) (r []byte) {
+	t := self.getTree()
```
this is not needed if Reset takes case of reserving a tree from the pool and assigning to `self.bmt`
same with ReadFrom

jmozah replied:
Well getTree doesnt Reserve twice.. it returns without doing anything if self.bmt is != nil

zelig commented on bmt/bmt.go:
```diff
@@ -0,0 +1,565 @@
+// reads from io.Reader and appends to the data to hash using Write
+// it reads so that chunk to hash is maximum length or reader reaches if EOF
+func (self *BMTHasher) ReadFrom(r io.Reader) (m int64, err error) {
+	self.getTree()
```
this is not needed if Reset takes case of reserving a tree from the pool and assigning to `self.bmt`
also in Sum

zelig commented on bmt/bmt.go:
```diff
@@ -0,0 +1,565 @@
+	return <-self.result
+}
+
+func (self *BMTHasher) Init(i int) {
```
this could be removed if we require `Reset`

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 3c19fac765a50b8f0746a8e4166c28f5732a0af3
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 0721141170312a259fed8b0d1f392e4b851bc9c7
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

lmars commented on swarm/storage/pyramid.go:
```diff
@@ -73,128 +130,496 @@ func NewPyramidChunker(params *ChunkerParams) (self *PyramidChunker) {
+		if storageWG != nil {
+			storageWG.Wait()
+		}
+		close(errC)
```
Why do we close `errC` here? Do we know there is no error by here? Shouldn't `prepareChunks` be responsible for closing the channel, given it is the one that potentially writes an error to the channel?

jmozah replied:
We close errC here because this is passed down to 8 parallel routines underneath.  if a error happens in one of them.. that particular routine raises error... once they all complete successfully, the control comes back and we can safely close this here.

lmars replied:
Perhaps add this as a comment?

jmozah replied:
Sure

lmars commented on swarm/storage/pyramid.go:
```diff
@@ -73,128 +130,496 @@ func NewPyramidChunker(params *ChunkerParams) (self *PyramidChunker) {
+		if err != nil {
+			return nil, err
+		}
+	case <-time.NewTimer(splitTimeout).C:
```
Same as the tree chunker, should an error be returned on timeout?

jmozah replied:
same answer as the tree chunker.

lmars commented on swarm/storage/pyramid.go:
```diff
@@ -73,128 +130,496 @@ func NewPyramidChunker(params *ChunkerParams) (self *PyramidChunker) {
+	// send off new chunk to storage
+	if chunkC != nil {
+		if swg != nil {
+			swg.Add(1)
```
If we are sending chunks to a channel here, why do we also have to increment a "storage" wait group? Shouldn't it be the receiver of the channel that is actually going to store the chunk who should also manage this wait group?

jmozah replied:
- The receiver function "storeWorker" in dpa uses only the channel to store the chunk.
- The swg in the above snippet is the chunk's own wg, which signals that the chunk is stored. This is used in testing to simulate a put.

The actual use of this is while storing in memstore, it is incremented and after it is stored in DB, it is decremented indicating that the operation is done. per chunk vise.

lmars commented on swarm/storage/pyramid.go:
```diff
@@ -73,128 +130,496 @@ func NewPyramidChunker(params *ChunkerParams) (self *PyramidChunker) {
+func (self *PyramidChunker) prepareChunks(isAppend bool, chunkLevel [][]*TreeEntry, data io.Reader, rootKey []byte, quitC chan bool, wg *sync.WaitGroup, jobC chan *chunkJob, processorWG *sync.WaitGroup, chunkC chan *Chunk, errC chan error, storageWG *sync.WaitGroup) {
+	defer wg.Done()
+
+	chunkWG := &sync.WaitGroup{}
```
What is this wait group used for?

jmozah replied:
This is where the parent  node waits for all the children nodes (128 branches) to complete OR for the input to get exhausted. Only when all the 128 hashes of the child nodes are processed.. the parent chunk can be constructed.

lmars commented on swarm/storage/pyramid.go:
```diff
@@ -73,128 +130,496 @@ func NewPyramidChunker(params *ChunkerParams) (self *PyramidChunker) {
+
+func (self *PyramidChunker) enqueueTreeChunk(chunkLevel [][]*TreeEntry, ent *TreeEntry, chunkWG *sync.WaitGroup, jobC chan *chunkJob, quitC chan bool, last bool) {
+
+	if ent != nil {
```
How about just not calling the function if this is `nil`?

jmozah replied:
enqueueTreeChunk is called in multiple places.. instead of putting the condition in multiple places.. i though it would be nice to have it in one place. :-)

lmars commented on swarm/storage/pyramid.go:
```diff
@@ -73,128 +130,496 @@ func NewPyramidChunker(params *ChunkerParams) (self *PyramidChunker) {
+
+}
+
+func (self *PyramidChunker) getHash(b []byte) {
```
Remove this empty method.

jmozah replied:
Done

jmozah commented on swarm/storage/chunker.go:
```diff
@@ -157,10 +142,15 @@ func (self *TreeChunker) Split(data io.Reader, size int64, chunkC chan *Chunk, s
+			return nil, err
+		}
+	case <-time.NewTimer(splitTimeout).C:
+
```
Right now it is a timeout exception to avoid routine leaks.... 
I plan to bring in Context for chunker in general while doing the "Concurrency" task.

lmars replied:
Ok at the very least it should log something then to help us identify such leaks (but I really think it should return an error).

jmozah replied:
done. returning errOperationTimedOut

jmozah commented on swarm/storage/chunker.go:
```diff
@@ -237,6 +229,9 @@ func (self *TreeChunker) split(depth int, treeSize int64, key Key, data io.Reade
 }
 
 func (self *TreeChunker) hashWorker(jobC chan *hashJob, chunkC chan *Chunk, errC chan error, quitC chan bool, swg, wwg *sync.WaitGroup) {
+
```
I removed them now.. Does gofmt allows these leading blank lines?

jmozah commented on swarm/storage/chunker_test.go:
```diff
@@ -21,8 +21,8 @@ import (
 	"crypto/rand"
 	"encoding/binary"
 	"fmt"
+	"github.com/ethereum/go-ethereum/crypto/sha3"
```
Done

jmozah commented on swarm/storage/chunker_test.go:
```diff
@@ -86,6 +86,53 @@ func (self *chunkerTester) Split(chunker Splitter, data io.Reader, size int64, c
+func (self *chunkerTester) Append(chunker Splitter, rootKey Key, data io.Reader, chunkC chan *Chunk, swg *sync.WaitGroup, expectedError error) (key Key) {
+
+	quitC := make(chan bool)
+	timeout := time.After(600 * time.Second)
```
Well.. i just reused this from the existing tests. i dont see a need for 600. I reduced it to 60 which is a good number for benchmarks that i run for large files.

jmozah commented on swarm/storage/chunker_test.go:
```diff
@@ -86,6 +86,53 @@ func (self *chunkerTester) Split(chunker Splitter, data io.Reader, size int64, c
+				case <-quitC:
+					return
+				case chunk := <-chunkC:
+					if chunk != nil {
```
This is not for channel closing.. This is where we simulate a request for chunk from the network. The chunk structure is supposed to be valid... this is just a guard check... before we access the values in the object.

jmozah commented on swarm/storage/chunker_test.go:
```diff
@@ -176,29 +223,148 @@ func testRandomData(splitter Splitter, n int, tester *chunkerTester) {
+	reader := tester.Join(chunker, newKey, 0, chunkC, quitC)
+	newOutput := make([]byte, n+m)
+	r, err := reader.Read(newOutput)
+	if r != (n+m) || err != io.EOF {
```
Yes.. i dont remember why i put this :-)

jmozah commented on swarm/storage/chunker_test.go:
```diff
@@ -176,29 +223,148 @@ func testRandomData(splitter Splitter, n int, tester *chunkerTester) {
+	}
+
+	newInput := append(input, appendInput...)
+	if newInput != nil {
```
Fixed.

jmozah commented on swarm/storage/chunker_test.go:
```diff
@@ -176,29 +223,148 @@ func testRandomData(splitter Splitter, n int, tester *chunkerTester) {
+	}
+
+	close(chunkC)
+	<-quitC
```
removed

jmozah commented on swarm/storage/chunker_test.go:
```diff
@@ -176,29 +223,148 @@ func testRandomData(splitter Splitter, n int, tester *chunkerTester) {
+	io.LimitReader(bytes.NewReader(input[8:]), int64(size))
+
+	rawSha3 := sha3.NewKeccak256()
+	rawSha3.Reset()
```
oh we have to. It initialises the hasher state in Reset.

zelig replied:
well, for this case of sha3 you dont need it

jmozah replied:
@zelig In sha3 too Reset does something with the permutation state and initializes some data structures.

jmozah commented on swarm/storage/chunker_test.go:
```diff
@@ -176,29 +223,148 @@ func testRandomData(splitter Splitter, n int, tester *chunkerTester) {
+	}
+
+	for i := range rawSha3Output {
+		if rawSha3Output[i] != sha3FromMakeFuncOutput[i] {
```
Fixed.

jmozah commented on swarm/storage/pyramid.go:
```diff
@@ -18,53 +18,110 @@ package storage
+	splitTimeout          = time.Minute * 5
+
+	DataChunk = iota
+	TreeChunk = iota
```
Yes.. i wanted to assign a sequence.. DataChunk as 0 and TreeChunk as 1.

lmars replied:
I personally think you should just assign them explicitly:

```go
const(
	DataChunk = 0
	TreeChunk = 1
)
```

jmozah replied:
Accepting this in favour of readability

jmozah commented on swarm/storage/pyramid.go:
```diff
@@ -73,128 +130,496 @@ func NewPyramidChunker(params *ChunkerParams) (self *PyramidChunker) {
+func (self *PyramidChunker) Split(data io.Reader, size int64, chunkC chan *Chunk, storageWG, processorWG *sync.WaitGroup) (Key, error) {
+
+	jobC := make(chan *chunkJob, 2*ChunkProcessors)
+	wg := &sync.WaitGroup{}
```
storageWG and proecessorWG are passed on by the dpa.
wg is the onlyone used in chunker.

storageWG: waits for DPA to complete the store call
proecessorWG: almost never used.. we should remove this in "concurrency" task i am going to do next
wg: This is used in chunker to wait for the processing threads to complete.

jmozah commented on swarm/fuse/fuse_file.go:
```diff
@@ -86,13 +101,29 @@ func (file *SwarmFile) Attr(ctx context.Context, a *fuse.Attr) error {
+
+	if file.fileSize == -1 {
+		reader := file.mountInfo.swarmApi.Retrieve(file.key)
+		quitC := make(chan bool)
```
removed

jmozah replied:
i think it is good to have the fuse related stuff in a separate PR.. I have removed it.

jmozah commented on swarm/storage/chunker.go:
```diff
@@ -223,11 +213,13 @@ func (self *TreeChunker) split(depth int, treeSize int64, key Key, data io.Reade
 			wwg.Add(1)
 		}
-		self.workerCount++
+		atomic.AddInt64(&self.workerCount, 1)
```
yeap.. added a mutex.

jmozah commented on swarm/storage/chunker_test.go:
```diff
@@ -86,6 +86,53 @@ func (self *chunkerTester) Split(chunker Splitter, data io.Reader, size int64, c
+			for {
+				select {
+				case <-timeout:
+					self.t.Fatalf("Join timeout error")
```
Done.
